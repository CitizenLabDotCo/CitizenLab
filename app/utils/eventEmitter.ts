import * as Rx from 'rxjs/Rx';

interface IEventEmitterEvent<T> {
  eventSource: string;
  eventName: string;
  eventValue: T;
}

class EventEmitter {
  private subject: Rx.Subject<IEventEmitterEvent<any>>;
  private stream: { [key: string]: Rx.Observable<IEventEmitterEvent<any>> };

  constructor() {
    this.subject = new Rx.Subject();
    this.stream = {};
  }

  emit<T>(eventSource: string, eventName: string, eventValue: T) {
    this.subject.next({ eventSource, eventName, eventValue });
  }

  observeEventFromSource<T>(eventSource: string, eventName: string): Rx.Observable<IEventEmitterEvent<T>> {
    const streamName = `${eventSource}-${eventName}`;

    if (!this.stream[streamName]) {
      this.stream[streamName] = this.subject.filter(data => data.eventSource === eventSource && data.eventName === eventName).share();
    }

    return this.stream[streamName];
  }

  observeEvent<T>(eventName: string): Rx.Observable<IEventEmitterEvent<T>> {
    const streamName = eventName;

    if (!this.stream[streamName]) {
      this.stream[streamName] = this.subject.filter(data => data.eventName === eventName).share();
    }

    return this.stream[streamName];
  }
}

const eventEmitter = new EventEmitter();
export default eventEmitter;
