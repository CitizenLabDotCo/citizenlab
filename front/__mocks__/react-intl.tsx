import { createIntl, MessageDescriptor } from 'react-intl';
import messages from '../app/i18n/en';
const Intl = jest.requireActual('react-intl');

// Initialise the real provider so that we don't
// need to reimplement any internals
const defaultProps = {
  locale: 'en',
  defaultLocale: 'en',
};

// The exact same `intl` object the real code receives ;-)
const intl = createIntl({ locale: 'en', messages });

const formatMessageReplacement = (
  messageDescriptor: MessageDescriptor,
  values?: { [key: string]: string | number | boolean | Date } | undefined
) => {
  return intl.formatMessage(messageDescriptor, {
    ...(values || {}),
  });
};

const intlReplacement = {
  ...intl,
  formatMessage: formatMessageReplacement,
};

// Set displayName so that snapshots don't use "Uknown" as component name
Intl.FormattedDate.displayName = 'FormattedDate';
Intl.FormattedTime.displayName = 'FormattedTime';
Intl.FormattedNumber.displayName = 'FormattedNumber';
Intl.FormattedPlural.displayName = 'FormattedPlural';
Intl.FormattedMessage.displayName = 'FormattedMessage';

// Special hook for tests, real package does not export this
Intl.intl = intlReplacement;

module.exports = Intl;
