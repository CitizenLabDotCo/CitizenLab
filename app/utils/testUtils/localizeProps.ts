// i18n
import { getLocalized } from 'utils/i18n';
import { InjectedLocalized } from 'utils/localize';

// Typing
import { Multiloc } from 'typings';

// A mock object to replace injectLocalize function : pass this to the child of injectLocalize as a prop
// ex : injectLocalize(TestedComponent) =>
// <TestedComponent {...localizeProps}>
export const localizeProps = {
  localize: (multiloc: Multiloc) => {
    return getLocalized(multiloc, 'en', ['en', 'fr-BE']);
  },
  locale: 'en',
  tenantLocales: ['en', 'fr-BE']
} as InjectedLocalized;
