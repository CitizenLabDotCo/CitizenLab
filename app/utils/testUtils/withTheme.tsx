import React from 'react';
import { ThemeProvider } from 'styled-components';
import { colors, fontSizes, stylingConsts } from 'utils/styleUtils';
import { shallow, mount } from 'enzyme';

export const theme = {
  colors,
  fontSizes,
  colorMain: '#ef0071',
  colorText: '#e68f51',
  colorSecondary: '#f76901',
  menuStyle: 'light',
  menuHeight: 74,
  mobileMenuHeight: 72,
  mobileTopBarHeight: 66,
  maxPageWidth: 952,
  ...stylingConsts
};

// export function shallowWithTheme<C extends Component, P = C['props'], S = C['state']>(tree: ReactElement<P>, options?: ShallowRendererProps) {
//   const context = shallow(<ThemeProvider theme={theme} />)
//     .instance()
//     .getChildContext();
//   return shallow(tree, {
//     ...options,
//     context: {
//       ...context,
//       intl
//     }
//   }) as ShallowWrapper<P, S, C>;
// }
//
export const shallowWithTheme = (tree) => {
  const context = (shallow(<ThemeProvider theme={theme} />).instance() as any).getChildContext();
  return shallow(tree, { context });
};

export const mountWithTheme = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    // @ts-ignore: TODO type this well
    .getChildContext();
  return mount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};
//
// export const mountWithTheme = (children: any) => mount(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
//
// export const shallowWithTheme = (children: any) => shallow(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
