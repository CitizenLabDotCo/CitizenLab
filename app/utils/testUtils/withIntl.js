import { IntlProvider, intlShape } from "react-intl";
import { shallow, mount } from "enzyme";
import React from 'react';

// Create IntlProvider to retrieve React Intl context
const intlProvider = new IntlProvider(
  {
    locale: "en",
    messages: {
      message1: "Hello world"
    }
  },
  {}
);
const { intl } = intlProvider.getChildContext();

// `intl` prop is required when using injectIntl HOC
const nodeWithIntlProp = node => React.cloneElement(node, { intl });

// shallow() with React Intl context
export const shallowWithIntl = (node, { context, ...options } = {}) => {
  return shallow(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    }
  });
};

// mount() with React Intl context
export const mountWithIntl = (
  node,
  { context, childContextTypes, ...options } = {}
) => {
  return mount(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    },
    childContextTypes: {
      intl: intlShape,
      ...childContextTypes
    }
  });
};
