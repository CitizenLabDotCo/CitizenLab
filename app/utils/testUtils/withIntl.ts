import { IntlProvider, intlShape, InjectedIntlProps } from 'react-intl';
import { shallow, mount, ShallowRendererProps, ShallowWrapper } from 'enzyme';
import React from 'react';

// Create IntlProvider to retrieve React Intl context
const intlProvider = new IntlProvider({
  locale: 'en',
}, {});

// You customize the intl object here:
const { intl: originalIntl } = intlProvider.getChildContext() as InjectedIntlProps;
const intl = {
  ...originalIntl,
  formatMessage: ({ id, defaultMessage }, values?) =>
    originalIntl.formatMessage({
      id,
      defaultMessage: defaultMessage || id
    }, { tenantName: 'The Test', orgName: 'Test Town', orgType: 'testing', ...values || {} }),
};

const nodeWithIntlProp = (node) => React.cloneElement(node, { intl });

// shallow() with React Intl context
export function shallowWithIntl<C extends React.Component, P = C['props'], S = C['state']>(node: React.ReactElement<P>, additional: ShallowRendererProps = {}) {
 const { context, ...options } = additional;
  return shallow(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    }
  }) as ShallowWrapper<P & InjectedIntlProps, S, C>;
}

// mount() with React Intl context
export const mountWithIntl = (
  node,
  { context, childContextTypes, ...options } = {}
) => {
  return mount(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    },
    childContextTypes: {
      intl: intlShape,
      ...childContextTypes
    }
  });
};
