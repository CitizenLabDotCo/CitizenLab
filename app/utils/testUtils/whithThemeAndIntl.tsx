// To render components with both intl and theme
// Never used yet, not sure how well it works
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { colors, fontSizes } from '../styleUtils';

import { shallow } from 'enzyme';

import { shallowWithIntl, mountWithIntl } from 'utils/testUtils/withIntl';
const theme = {
  colors,
  fontSizes,
  colorMain: '#ef0071',
  menuStyle: 'light',
  menuHeight: 74,
  mobileMenuHeight: 72,
  mobileTopBarHeight: 66,
  maxPageWidth: 952,
};

export const shallowWithThemeIntl = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallowWithIntl(tree, { context });
};
export const mountWithThemeIntl = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return mountWithIntl(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};
