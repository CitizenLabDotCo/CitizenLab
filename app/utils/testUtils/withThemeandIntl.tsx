import React from 'react';
import { ThemeProvider } from 'styled-components';
import { colors, fontSizes } from '../styleUtils';

import { shallow, mount } from 'enzyme';

import { shallowWithIntl, mountWithIntl } from 'utils/testUtils/withIntl';

const theme = {
  colors,
  fontSizes,
  colorMain: '#ef0071',
  menuStyle: 'light',
  menuHeight: 74,
  mobileMenuHeight: 72,
  mobileTopBarHeight: 66,
  maxPageWidth: 952,
};

export const shallowWithHoCs = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallowWithIntl(tree, { context });
};
export const mountWithHoCs = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return mountWithIntl(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};
