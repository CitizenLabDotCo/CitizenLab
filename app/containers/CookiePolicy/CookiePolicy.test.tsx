import React from 'react';
import 'jest-styled-components';
jest.mock('utils/cl-intl');
jest.mock('services/pages');
jest.mock('utils/eventEmitter');
import { CookiePolicy } from './';
import eventEmitter from 'utils/eventEmitter';

// tslint:disable-next-line:no-vanilla-formatted-messages
import { FormattedMessage } from 'react-intl';

import { shallowWithIntl } from 'utils/testUtils/withIntl';
import { shallow } from 'enzyme';

const emit = eventEmitter.emit as jest.Mock;
type FormattedMessageProps = FormattedMessage.Props;

describe('<CookiePolicy />', () => {
  it('renders correctly', () => {
    const wrapper = shallowWithIntl(<CookiePolicy />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly openConsentManager button', () => {
    const wrapper = shallowWithIntl(<CookiePolicy />);
    const cookieMessage = wrapper.find('FormattedMessage.cookiePreferencesMessage');
    const values = cookieMessage.props().values as FormattedMessageProps['values'];
    const prefbutton = values && values.changePreferencesButton as JSX.Element | null;
    const button = prefbutton && shallow(prefbutton);
    expect(button).toMatchSnapshot();
  });

  /* this is a pretty reliant on the implementation way to test that it opens consent manager.
    ideally, this would be tested by integration testing
  */
  it('has a functional openConsentManager Button', () => {
    // render component
    const wrapper = shallowWithIntl(<CookiePolicy />);
    expect(eventEmitter.emit).toBeCalledTimes(0);

    const cookieMessage = wrapper.find('FormattedMessage.cookiePreferencesMessage');
    const values = cookieMessage.props().values as FormattedMessageProps['values'];
    const prefbutton = values && values.changePreferencesButton as JSX.Element | null;
    const button = prefbutton && shallow(prefbutton);
    button && button.simulate('click');
    expect(eventEmitter.emit).toBeCalledTimes(1);
    expect(emit.mock.calls[0][1]).toBe('openConsentManager');
  });
});
