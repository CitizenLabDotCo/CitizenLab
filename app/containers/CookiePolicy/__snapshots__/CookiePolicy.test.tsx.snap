// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CookiePolicy /> renders correctly 1`] = `
<CookiePolicy__Container
  className="e2e-page-cookie-policy"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Cookie Policy
    </title>
    <meta
      content="A detailed explanation of how we use cookies on this platform"
      name="description"
    />
  </HelmetWrapper>
  <CookiePolicy__PageContent>
    <CookiePolicy__SContentContainer>
      <Component
        name="pages/cookie-policy/content"
      >
        <CookiePolicy__PageTitle>
          <FormattedMessage
            defaultMessage="Cookie Policy"
            id="app.containers.CookiePolicy.cookiePolicyTitle"
          />
        </CookiePolicy__PageTitle>
        <CookiePolicy__PageDescription>
          <QuillEditedContent>
            <FormattedMessage
              defaultMessage="As most websites, we use cookies to optimise the experience you and other visitors have on this platform. Because we want to be fully transparent on the why & how of this cookie use, you’ll find all details below, in as clearly written wording as possible. Cookies used on our platform are never used to identify and track specific users, they don’t \\"know who you are\\". Important to stress though, is that they do track technical data, such as a browser info, approximate location and an IP address. Although they are not used in such a way, in combination with other data sources, this could lead to an identification."
              id="app.containers.CookiePolicy.intro"
              tagName="p"
            />
            <FormattedMessage
              className="cookiePreferencesMessage"
              defaultMessage="At any time, you can change your preferences {changePreferencesButton}."
              id="app.containers.CookiePolicy.changePreferences"
              tagName="p"
              values={
                Object {
                  "changePreferencesButton": <CookiePolicy__StyledButton
                    onClick={[Function]}
                  >
                    <FormattedMessage
                      defaultMessage="here"
                      id="app.containers.CookiePolicy.changePreferencesButtonText"
                    />
                  </CookiePolicy__StyledButton>,
                }
              }
            />
            <FormattedMessage
              defaultMessage="Who are we?"
              id="app.containers.CookiePolicy.whoAreWeTitle"
              tagName="h2"
            />
            <FormattedMessage
              defaultMessage="We are {citizenLabLink}, a privately owned Belgian company. As software provider, we offer a digital participation platform to (local) governments and organisations. An example of such a platform is the one you’re currently on."
              id="app.containers.CookiePolicy.whoAreWeContent"
              tagName="p"
              values={
                Object {
                  "citizenLabLink": <a
                    href="https://www.citizenlab.co"
                    target="_blank"
                  >
                    CitizenLab
                  </a>,
                }
              }
            />
            <FormattedMessage
              defaultMessage="What are cookies?"
              id="app.containers.CookiePolicy.whatAreCookiesTitle"
              tagName="h2"
            />
            <FormattedMessage
              defaultMessage="Cookies are little text files which are stored on the browser or on the hard drive of your computer or mobile device when you visit a webpage or application. Cookies remember your actions and preferences. Logged in last time? A cookie will make sure you’re automatically logged in the next time you visit us. Visited the platform two weeks ago? A cookie will prove helpful to highlight what has been added to the platform in the last two weeks. More info on cookies in general, can be found in {wikipediaCookieLink}."
              id="app.containers.CookiePolicy.whatAreCookiesContent"
              tagName="p"
              values={
                Object {
                  "wikipediaCookieLink": <a
                    href="https://en.wikipedia.org/wiki/HTTP_cookie"
                    target="_blank"
                  >
                    this wikepedia artice
                  </a>,
                }
              }
            />
            <FormattedMessage
              defaultMessage="What cookies do we use?"
              id="app.containers.CookiePolicy.whatCookiesTitle"
              tagName="h2"
            />
            <FormattedMessage
              defaultMessage="Several third party tools are used on each of our participation platforms. These tools in turn use cookies to be able to function."
              id="app.containers.CookiePolicy.whatCookiesContent"
            />
            <FormattedMessage
              defaultMessage="Analytics"
              id="app.containers.CookiePolicy.analyticsTitle"
              tagName="h3"
            />
            <FormattedMessage
              defaultMessage="We only use mass analytics. Meaning that no individual data are used, and that you’re browsing our platform anonymously. These aggregated data, or mass analytics, show us which pages are the most and least popular and see how visitors move around the platform. This is necessary information for continuous improvement of the user experience. {analyticsLink} is our primary tool for analytics."
              id="app.containers.CookiePolicy.analyticsContent"
              tagName="p"
              values={
                Object {
                  "analyticsLink": <a
                    href="https://en.wikipedia.org/wiki/Google_Analytics"
                    target="_blank"
                  >
                    Google Analytics
                  </a>,
                }
              }
            />
            <FormattedMessage
              defaultMessage="Advertising"
              id="app.containers.CookiePolicy.advertisingTitle"
              tagName="h3"
            />
            <FormattedMessage
              defaultMessage="As CitizenLab, we don’t use advertising tools on the participation platforms, but some (local) governments and organisations do. {advertisingLink} is the most used tool for this. This allows them to learn what type of users took certain actions, and to build Facebook advertising campaigns based on that knowledge."
              id="app.containers.CookiePolicy.advertisingContent"
              tagName="p"
              values={
                Object {
                  "advertisingLink": <a
                    href="https://www.facebook.com/business/help/742478679120153?helpref=faq_content"
                    target="_blank"
                  >
                    Facebook Pixel
                  </a>,
                }
              }
            />
            <FormattedMessage
              defaultMessage="Functional"
              id="app.containers.CookiePolicy.functionalTitle"
              tagName="h3"
            />
            <FormattedMessage
              defaultMessage="The main functional tool we use in this regard is {functionalLink}. This tool allows us to talk to you in a direct manner when visiting the platform. This way, we could show you around, ask for feedback on the platform… The cookie allows us to continue this conversation when you come back. Off course, still without knowing who you are. When logged in, we use them to be able to send you in-app and email notifications on certain events relevant to you. These notification settings can be changed at any time on your user profile page."
              id="app.containers.CookiePolicy.functionalContent"
              tagName="p"
              values={
                Object {
                  "functionalLink": <a
                    href="https://www.intercom.com/"
                    target="_blank"
                  >
                    Intercom
                  </a>,
                }
              }
            />
            <FormattedMessage
              defaultMessage="You can find a full, up-to date list of 3rd parties and manage your preferences {cookiesListButton}."
              id="app.containers.CookiePolicy.cookiesList"
              tagName="p"
              values={
                Object {
                  "cookiesListButton": <CookiePolicy__StyledButton
                    className="cookieList"
                    onClick={[Function]}
                  >
                    here
                  </CookiePolicy__StyledButton>,
                }
              }
            />
            <FormattedMessage
              defaultMessage="Further questions or remarks? You'll find us here: support@citizenlab.co."
              id="app.containers.CookiePolicy.contactInfo"
              tagName="p"
            />
          </QuillEditedContent>
        </CookiePolicy__PageDescription>
      </Component>
    </CookiePolicy__SContentContainer>
  </CookiePolicy__PageContent>
  <CookiePolicy__PagesNavWrapper>
    <CookiePolicy__PagesNav>
      <CookiePolicy__SContentContainer>
        <CookiePolicy__StyledLink
          key="information"
          to="/pages/information"
        >
          <FormattedMessage
            defaultMessage="Information Page"
            id="app.containers.CookiePolicy.informationPageName"
          />
          <CookiePolicy__LinkIcon
            name="chevron-right"
          />
        </CookiePolicy__StyledLink>
        <CookiePolicy__StyledLink
          key="terms-and-conditions"
          to="/pages/terms-and-conditions"
        >
          <FormattedMessage
            defaultMessage="Terms & conditions"
            id="app.containers.CookiePolicy.terms-and-conditionsPageName"
          />
          <CookiePolicy__LinkIcon
            name="chevron-right"
          />
        </CookiePolicy__StyledLink>
        <CookiePolicy__StyledLink
          key="privacy-policy"
          to="/pages/privacy-policy"
        >
          <FormattedMessage
            defaultMessage="Privacy Policy"
            id="app.containers.CookiePolicy.privacy-policyPageName"
          />
          <CookiePolicy__LinkIcon
            name="chevron-right"
          />
        </CookiePolicy__StyledLink>
      </CookiePolicy__SContentContainer>
    </CookiePolicy__PagesNav>
  </CookiePolicy__PagesNavWrapper>
  <WrappedFooter
    showCityLogoSection={false}
  />
</CookiePolicy__Container>
`;

exports[`<CookiePolicy /> renders correctly openConsentManager button 1`] = `
.c0 {
  color: #147985;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  margin: 0;
  padding: 0;
  cursor: pointer;
}

.c0:hover {
  color: #0a3d43;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<button
  className="c0"
  onClick={[Function]}
>
  <FormattedMessage
    defaultMessage="here"
    id="app.containers.CookiePolicy.changePreferencesButtonText"
  />
</button>
`;
